packages/gatsby-plugin-sharp/src/gatsby-node.js

```diff
@@ -1,10 +1,84 @@
-const { setBoundActionCreators, setPluginOptions } = require(`./index`)
+const {
+  setBoundActionCreators,
+  setPluginOptions,
+  queue,
+  reportError,
+} = require(`./index`)
+const processFile = require(`./processFile`)
+const { scheduleJob } = require(`./scheduler`)
 
-exports.onPreInit = ({ actions }, pluginOptions) => {
+const getQueueFromCache = async cache => {
+  const rawQueue = await cache.get(`queue`)
+
+  if (!rawQueue) {
+    return new Map()
+  }
+
+  return new Map(rawQueue)
+}
+
+exports.onPreInit = async ({ actions }, pluginOptions) => {
   setBoundActionCreators(actions)
   setPluginOptions(pluginOptions)
 }
 
+/**
+ * save queue to the cache
+ */
+exports.onPostBootstrap = async ({ cache }) => {
+  const cachedQueue = await getQueueFromCache(cache)
+
+  // merge both queues
+  for (const [key, job] of cachedQueue) {
+    queue.set(key, job)
+  }
+
+  // JSON.stringify doesn't work on an Map so we need to convert it to an array
+  return cache.set(`queue`, Array.from(queue))
+}
+
+/**
+ * Execute all unprocessed images on gatsby build
+ */
+let promises = []
+exports.onPreBuild = async ({ cache, boundActionCreators }, pluginOptions) => {
+  const cachedQueue = await getQueueFromCache(cache)
+
+  let promises = []
+  for (const [, job] of cachedQueue) {
+    promises.push(scheduleJob(job, boundActionCreators, pluginOptions))
+  }
+}
+
+/**
+ * wait for all images to be processed
+ */
+exports.onPostBuild = ({ cache, reporter }) =>
+  Promise.all(promises)
+    .then(() => cache.set(`queue`, []))
+    .catch(({ err, message }) => {
+      reportError(message || err.message, err, reporter)
+    })
+
+/**
+ * Build images on the fly when they are requested by the browser
+ */
+exports.onCreateDevServer = ({ app, cache }, pluginOptions) => {
```

Comment:
This is slick!

Reply:
nifty!

Could we name this `onCreateDevelopmentServer`? I'm strongly allergic to shorten names for APIs :-)

Is the changes to develop.js not pushed yet? Plus the API docs changes?

Reply:
Oh, this is not new API (we have `onCreateDevServer` API for a while). So this is my bad for adding this API with shorten name :( mea culpa